### PR TITLE
Refactor permutation two sided

### PR DESCRIPTION
**Pull request by Ali-Tehrani**
This pull request is regarding permutation for two-sided, specifically the function permutation_2sided

The variables for unpad_col, unpad_row and pad were updated to match the other functions.
For single transformation, a check was added to make sure that they are square matrices.
Certain part of the code was moved into the if statement.

### CODE REVIEW

File: procrustes/permutation.py [hunk 1]
```diff
@@ -371,25 +370,32 @@ def permutation_2sided(array_a, array_b, transform_mode="single",
...
     transform_mode = transform_mode.lower()
     if transform_mode == "single":
+        # np.power() can not handle the negatives values
+        # Try to convert the matrices to non-negative
+        # shift the the matrices to avoid negative values
+        # otherwise it will cause an error in the Eq. 28 in the research notes
+        maximum = max(np.amax(np.abs(new_a)), np.amax(np.abs(new_b)))
+        new_a_positive = new_a.astype(np.float) + maximum
+        new_b_positive = new_b.astype(np.float) + maximum
+
```
Comment (Ali-Tehrani) [inline procrustes/permutation.py:392]: I'm not sure why this part is needed to make sure that it is positive.  The comments aren't obvious to me why np.power can't handle negative values or what Eq. 28 is (maybe it is Eq B.29?). I was wondering if @fwmeng88 remembers?  If it is Eq B.29, it seems that it should be able to handle negative values.

**Comment by Ali-Tehrani** [comment]
Also for the add_noise argument , Its used when the transformation is single and when the matrices are symmetric.  
The document says it is to make sure that it is diagonalizable, but symmetric matrices are always diagonalizable, so I'm not certain if this argument is needed or why it was added.

**Comment by Ali-Tehrani** [comment]
For the single transformation, I think that the (symmetric matrices) undirected and (non-symmetric) directed can be grouped together because the technique used in Umeyama's paper (and the other paper) is to convert the non-symmetric matrix to a Hermitian matrix and then find a initial guess exactly like the (symmetric) undirected case.   However, the updating rules are different between the two.  Any thoughts @FarnazH, @PaulWAyers, @fwmeng88?

File: procrustes/permutation.py (hunk 1 above)
Comment (fwmeng88) [reply procrustes/permutation.py:392]: This comment corresponds to the Equation 28 in the notes. I will email you the notes shortly. The reason why we need to ensure the non-negativity is square root does not allow negative values. Eq 28 can be generalized to Eq. 8 and Eq. 14 in paper [Nonnegative matrix factorization for combinatorial optimization: Spectral clustering, graph matching, and clique finding](https://ieeexplore.ieee.org/abstract/document/4781113/?casa_token=GNr6vDvB6NoAAAAA:Jks0I4SEV8FG9Exn-5qL-wtCReYTmCvZBIXtshjC7wAwmQFR4WpO4iw-_y4Z95-kfo6r90Qeow), which is exactly what we implemented for directed and undirected.

If we comment on these lines of code, some tests will fail. Hope this helps a little. @Ali-Tehrani

Comment (Ali-Tehrani) [reply procrustes/permutation.py:392]: Thank you it's a lot more clear now.

**Comment by fwmeng88** [comment]
> Also for the add_noise argument , Its used when the transformation is single and when the matrices are symmetric.
> The document says it is to make sure that it is diagonalizable, but symmetric matrices are always diagonalizable, so I'm not certain if this argument is needed or the actual reason as to why it was initially added.

At that point, we don't require a user to ensure the input of a symmetric matrix but a squared one. So, I think this `add_noise` is implemented for this situation if I recalled correctly.

**Comment by fwmeng88** [comment]
> For the single transformation, I think that the (symmetric matrices) undirected and (non-symmetric) directed can be grouped together because the technique used in Umeyama's paper (and the other paper) is to convert the non-symmetric matrix to a Hermitian matrix and then find a initial permutation matrix guess exactly like the (symmetric) undirected case. So, finding the initial guess is the same between the two but the updating rules are different.
> Any thoughts @FarnazH, @PaulWAyers, @fwmeng88?

Thanks for noticing this! @Ali-Tehrani 

The undirected scenario is a special case of directed one. The updating details can be found at  [Nonnegative matrix factorization for combinatorial optimization: Spectral clustering, graph matching, and clique finding](https://ieeexplore.ieee.org/abstract/document/4781113/?casa_token=GNr6vDvB6NoAAAAA:Jks0I4SEV8FG9Exn-5qL-wtCReYTmCvZBIXtshjC7wAwmQFR4WpO4iw-_y4Z95-kfo6r90Qeow). I think we can do that.

File: procrustes/permutation.py [hunk 2]
```diff
@@ -136,12 +136,24 @@ def permutation(
...
+        translate=False,
+        scale=False,
+        mode="normal1",
+        check_finite=True,
+        iteration=500,
+        add_noise=False,
+        tol=1.0e-8,
+        kopt=False,
+        kopt_k=3,
+        weight=None
```
Comment (FarnazH) [inline procrustes/permutation.py:155]: This is missing from the docstring.

File: procrustes/permutation.py [hunk 3]
```diff
@@ -136,12 +136,24 @@ def permutation(
...
+        unpad_row=False,
+        translate=False,
+        scale=False,
+        mode="normal1",
+        check_finite=True,
+        iteration=500,
+        add_noise=False,
+        tol=1.0e-8,
+        kopt=False,
+        kopt_k=3,
```
Comment (FarnazH) [inline procrustes/permutation.py:154]: I would combine these and just have one argument. In other words, if `kopt=None`, the kop algorithm is skipped, However, if is given an integer value, the kopt step is added. Let me know if you have other suggestions.

File: procrustes/permutation.py [hunk 4]
```diff
@@ -136,12 +136,24 @@ def permutation(
...
-def permutation_2sided(array_a, array_b, transform_mode="single",
-                       remove_zero_col=True, remove_zero_row=True,
-                       pad_mode="row-col", translate=False, scale=False,
-                       mode="normal1", check_finite=True, iteration=500,
-                       add_noise=False, tol=1.0e-8, kopt=False, kopt_k=3,
-                       weight=None):
+def permutation_2sided(
+        array_a,
+        array_b,
+        transform_mode="single",
```
Comment (FarnazH) [inline procrustes/permutation.py:142]: This argument can also change to `single=True`, if `False` the double permutation Procrustes is performed.

File: procrustes/permutation.py [hunk 5]
```diff
@@ -136,12 +136,24 @@ def permutation(
...
 
-def permutation_2sided(array_a, array_b, transform_mode="single",
-                       remove_zero_col=True, remove_zero_row=True,
-                       pad_mode="row-col", translate=False, scale=False,
-                       mode="normal1", check_finite=True, iteration=500,
-                       add_noise=False, tol=1.0e-8, kopt=False, kopt_k=3,
-                       weight=None):
+def permutation_2sided(
+        array_a,
+        array_b,
```
Comment (FarnazH) [inline procrustes/permutation.py:141]: Please change these to `a` and `b` to be consistent with other functions & scipy/numpy.

File: procrustes/permutation.py (hunk 1 above)
Comment (FarnazH) [reply procrustes/permutation.py:392]: power can handle square rooting of negative values, it just results in a complex number that we are not covering at this stage.
Also, please clarify the comments as just mentioning Eq. 28 is not helpful.

**Comment by PaulWAyers** [comment]
> > Also for the add_noise argument , Its used when the transformation is single and when the matrices are symmetric.
> > The document says it is to make sure that it is diagonalizable, but symmetric matrices are always diagonalizable, so I'm not certain if this argument is needed or the actual reason as to why it was initially added.
> 
> At that point, we didn't require a user to ensure the input of a symmetric matrix but a squared one. So, I think this `add_noise` is implemented for this situation if I recalled correctly.

I think that it's fine to remove the add_noise argument now. It's no longer needed.

**Comment by PaulWAyers** [comment]
> For the single transformation, I think that the (symmetric matrices) undirected and (non-symmetric) directed can be grouped together because the technique used in Umeyama's paper (and the other paper) is to convert the non-symmetric matrix to a Hermitian matrix and then find a initial permutation matrix guess exactly like the (symmetric) undirected case. So, finding the initial guess is the same between the two but the updating rules are different.
> Any thoughts @FarnazH, @PaulWAyers, @fwmeng88?

I agree. There is a small gain in efficiency from treating the undirected case separately, but it is *much* easier for the user to just pass a matrix. Then we assume nothing, and use the directed-graph stategy in all cases. (I.e., only section 3.2 of [1]; section IV.A. of [2].) This seems easier for the user and costs us very little.....

Incidentally, the issue with needing to treat this as a graph-matching problem (ergo needing nonnegative matrices) may not be insurmountable. @fwmeng88 would know whether or not there is precedent/references for the |a_ij| trick. But it might be (I don't know) that the mathematics does not rely upon this assumption. I feel much more confident now, since the imaginary parts of the Hermitian matrix in [2] are generally *not* nonnegative. It seems, then, that the Umeyama method does not require nonnegative matrices. 

For the Ding-Li-Jordan method, it seems possible to square both sides of the update Eq. (14); that would remove the need for assumptions on the matrix elements. (Then P_{ij} would obviously be set equal to the positive square root of the P_{ij}^2 elements computed.)  It needs to be double-checked, but the update rule derived from the inline equation in the very last line in the proof of Theorem 3, which involves the square of the permutation matrix(!), would be precisely the "squared" update rule. 

I think, therefore, that we can (a) use Umeyama with or without assuming the matrices are nonnegative and (b) square the update rule in the  Ding-Li-Jordan method to remove the requirement of square matrices. I did not really go through all the math in [1] and [2], however.

[1] C. Ding, T. Li and M. I. Jordan, "Nonnegative Matrix Factorization for Combinatorial Optimization: Spectral Clustering, Graph Matching, and Clique Finding," 2008 Eighth IEEE International Conference on Data Mining, Pisa, Italy, 2008, pp. 183-192, doi: 10.1109/ICDM.2008.130.
[2] S. Umeyama.  An eigendecomposition approach toweighted graph matching problems.IEEE Trans. onPattern Analysis and Machine Intelligence, 10:695 –703, 1988.

File: procrustes/permutation.py (hunk 4 above)
Comment (Ali-Tehrani) [reply procrustes/permutation.py]: This is resolved in commit (677ee8a and 33ff563)

File: procrustes/permutation.py (hunk 5 above)
Comment (Ali-Tehrani) [reply procrustes/permutation.py]: This is resolved in 677ee8a and 33ff563

File: procrustes/permutation.py (hunk 3 above)
Comment (Ali-Tehrani) [reply procrustes/permutation.py]: This is resolved in bf8cc6f

File: procrustes/permutation.py (hunk 2 above)
Comment (Ali-Tehrani) [reply procrustes/permutation.py:160]: This is resolved in 98d45a9

**Comment by Ali-Tehrani** [comment]
> > > Also for the add_noise argument , Its used when the transformation is single and when the matrices are symmetric.
> > > The document says it is to make sure that it is diagonalizable, but symmetric matrices are always diagonalizable, so I'm not certain if this argument is needed or the actual reason as to why it was initially added.
> > 
> > 
> > At that point, we didn't require a user to ensure the input of a symmetric matrix but a squared one. So, I think this `add_noise` is implemented for this situation if I recalled correctly.
> 
> I think that it's fine to remove the add_noise argument now. It's no longer needed.

**Comment by Ali-Tehrani** [comment]
> Incidentally, the issue with needing to treat this as a graph-matching problem (ergo needing nonnegative matrices) may not be insurmountable. @fwmeng88 would know whether or not there is precedent/references for the |a_ij| trick. But it might be (I don't know) that the mathematics does not rely upon this assumption. I feel much more confident now, since the imaginary parts of the Hermitian matrix in [2] are generally _not_ nonnegative. It seems, then, that the Umeyama method does not require nonnegative matrices.

The way I understood it is that the objective function ||P^T A P - B|| doesn't change when adding a constant value to both A and B, because the permutation of a matrix with identical entries is itself. 

> For the Ding-Li-Jordan method, it seems possible to square both sides of the update Eq. (14); that would remove the need for assumptions on the matrix elements. (Then P_{ij} would obviously be set equal to the positive square root of the P_{ij}^2 elements computed.) It needs to be double-checked, but the update rule derived from the inline equation in the very last line in the proof of Theorem 3, which involves the square of the permutation matrix(!), would be precisely the "squared" update rule.

I tried playing around with this (code wise) and didn't have much success. I think it has to do with the square root of the iterative "permutation" matrix must be taken regardless, which can have negative values. One could take the absolute of this "iterative" permutation matrix but given that constant-value translation is invariant, I decided to leave it.

What I meant to say was about the undirected (symmetric) case having 4 initial guess  whereas the directed (unsymmetric) case has only one initial guess (Umeyama one from [1], [2]).  I know the other Umeyama-variant guess can also be brought over however the two-other initial guess, I don't understand it and one has to deal with complex numbers so I decided not to touch it and backtrack on this suggestion.  One could raise an issue regarding this.

I don't have anything else that I can add to this pull request.

**Review by fwmeng88 — APPROVED** [review]

**Comment by fwmeng88** [comment]
Thank you for helping! Looks nice to me and I will merge this PR for now. 
@Ali-Tehrani

**Comment by PaulWAyers** [comment]
It's a good point that adding a constant (the same constant) to both matrices doesn't change the solution. That's a very good strategy then, and preferred. I'd just go with that (which I think is what we already do). We should point it out clearly if we haven't already.

**Comment by fwmeng88** [comment]
> It's a good point that adding a constant (the same constant) to both matrices doesn't change the solution. That's a very good strategy then, and preferred. I'd just go with that (which I think is what we already do). We should point it out clearly if we haven't already.

We already did this, by adding both matrices with a maximum value of elements in two matrices. https://github.com/theochem/procrustes/blob/47d1f58810e3c636cd468baa5b878d39dfa987d3/procrustes/permutation.py#L401-L403.

**Pull request merged**